compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -713,6 +714,56 @@ impl<'a> Parser<'a> {
         }
     }
 
+    /// Check if a method call or field access was written trying to interpolate with a macro
+    /// substitution. For example, `foo.bar_$baz()`.
+    pub(super) fn check_bad_macro_interpolation(
+        &mut self,
+        segment: &PathSegment,
+    ) -> Result<(), PResult<'a, P<Expr>>> {
+        if let token::Interpolated(nonterminal) = &self.token.kind {
+            if let (token::NtIdent(..), true) = (
+                nonterminal.deref(),
+                self.look_ahead(1, |t| matches!(t.kind, token::OpenDelim(_) | token::Semi)),
+            ) {
+                // Complain about trying to interpolate a field name.
+                let msg = "unexpected macro fragment";
+                let span = self.token.span;
+                let token_str = pprust::token_kind_to_string(&self.token.kind);
+                let mut err = self.struct_span_err(span, &format!("{}: {}", msg, token_str));
+                err.span_label(span, msg);
+                if self.sess.unstable_features.is_nightly_build() {
+                    err.multipart_suggestion(
+                        "the nightly only `std::concat_idents` macro can be used to synthesize \
+                         identifiers",
+                        vec![
+                            (segment.ident.span.shrink_to_lo(), "concat_idents!(".to_string()),
+                            (segment.ident.span.shrink_to_hi().until(span), ", ".to_string()),
+                            (span.shrink_to_hi(), ")".to_string()),
+                        ],
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+                err.multipart_suggestion(
+                    "the `paste` crate can be used to synthesize identifiers in macros",
+                    vec![
+                        (segment.ident.span.shrink_to_lo(), "paste! { [<".to_string()),
```

Comment:
I like this suggestion, but the current fixup is probably quite confusing to newcomers who don't understand that macros don't work in identifer position. 

Is there a way of grabbing the outermost expression span and suggest wrapping `paste!` around that instead? I guess this might be difficult because we don't know the whole expr yet.

Also perhaps consuming as many ident/macro fragments until we see `::` or `(` so we don't suggest:
```diff
- a $b c
+ paste!([<a $b>]) c
but instead
+ paste!([a $b c])
```

Otherwise, maybe it's better to just use span_note suggesting use of these macros.